### DISCUSSION TITLE
【新提案】四个新一元运算符 await.all / await.race / await.allSettled / await.any

### DISCUSSION THREAD

**Discussion by Jack-Works** [JSCIG/es-discuss#12] [Stage 1]
我觉得这个想法很有意思，能简化批量处理 Promise 的用例（是 ljharb 提的 idea）。不知道有没有人有兴趣一起把它变成提案？

请参阅：https://es.discourse.group/t/allow-awaiting-on-arrays/178/19

**Comment by ljharb** [comment]
I'd be happy to co-champion it if someone wanted to invest the time to do the legwork (my plate is a bit too full to take that on right now).

**Comment by Jack-Works** [comment]
https://github.com/Jack-Works/proposal-await.ops

**Comment by hax** [comment]
这个东西我精神上是支持的。只是它的motivation仅仅是少打一点字符，略显单薄。最好能再加点啥。比如或许引擎可以更好地优化啥的（虽然感觉也就是个车轱辘话）。

**Comment by Jack-Works** [comment]
我觉得引擎在这里没什么好优化的，除了能省下去全局找 Promise.* 方法的时间。
的确就是少打几个字这么简单了，所以如果不过的话也无所谓（反正天知道要等多久才能用上）

**Comment by ljharb** [comment]
For me, the motivation isn't just that it's shorter or that it's more robust (avoiding the dynamic `Promise` lookup) - just like `await` actually allows engines to provide better stack traces than they have for `.then`, `await.all` and friends should theoretically allow the same improvements.

**Comment by Jack-Works** [comment]
<table><tbody><tr><td><code>await</code> actually allows engines to provide better stack traces than they have for <code>.then</code>, <code>await.all</code> and friends should theoretically allow the same improvements.</td></tr></tbody></table>

I didn't notice about stack traces before 🙈

**Comment by hax** [comment]
> allows engines to provide better stack traces

Great! can we add some description in README?

offtopic: @ljharb What translation service r u use to read chinese ? 😁

**Comment by Jack-Works** [comment]
![](https://user-images.githubusercontent.com/5390719/87216318-3c620900-c371-11ea-8152-5c33f9f5f131.png)

Current call stack. What's the better form of it?

**Comment by ljharb** [comment]
Google Translate :-) hopefully I’m getting a close approximation :-p

I’m not sure await provides a better call stack with Promise.all currently - check https://mathiasbynens.be/notes/async-stack-traces and https://cloudreports.net/v8-zero-cost-async-stack-traces/ i think?

**Comment by xiaoxiangmoe** [comment]
只是语法糖的话，感觉用处不大吧，就是少打了几个字。而且，会不会有一些人要求把各种函数都做成算符内置进语法呢？

**Comment by hax** [comment]
> 会不会有一些人要求把各种函数都做成算符内置进语法呢？

@xiaoxiangmoe 比如说？

**Comment by xiaoxiangmoe** [comment]
> > 会不会有一些人要求把各种函数都做成算符内置进语法呢？
> 
> @xiaoxiangmoe 比如说？

Promise 的几个函数能做算符，凭什么我 Array 的不行，凭什么我函数上的不行……于是就出现了很多算符……（这应该去推进自定义算符的提案）

**Comment by ljharb** [comment]
To me, since `await` already exists, there's already precedence for "operators as sugar over Promises", and this just provides the missing pieces.
